admin/aaxplay.1.in
.\" Manpage for aaxplay.
.\" Contact tech@adalin.com to correct errors or typos.
.TH man 1 "19 Dec 2014" "@AAX_UTILS_MAJOR_VERSION@.@AAX_UTILS_MINOR_VERSION@.@AAX_UTILS_MICRO_VERSION@" "aaxplay man page"
.SH NAME
aaxplay \- Plays audio from a file or from an audio input device.
.SH SYNOPSIS
.B aaxplay
[\fIOPTION\fR]...
.SH DESCRIPTION
.PP
Plays audio from a file or from an audio input device.
Optionally writes the audio also to an output file.
.TP
\fB\-i\fR, \fB\-\-input \fRFILE\fR
playback audio from a file
.TP
\fB\-i\fR, \fB\-\-capture \fRDEVICE\fR
capture from an audio device
.TP
\fB\-d\fR, \fB\-\-device \fRDEVICE\fR
playback device (default if not specified)
.TP
\fB\-o\fR, \fB\-\-output \fRFILE\fR
also write to an audio file (optional)
.TP
\fB\-b\fR, \fB\-\-batch
process as fast as possible (Audio Files device only)
.TP
\fB\-v\fR, \fB\-\-verbose
show extra playback information
.TP
\fB\-h\fR, \fB\-\-help
print this message and exit
.PP
Either --input or --capture can be used but not both.
.PP
For a list of device names run: aaxinfo
.PP
Audio will always be sent to the (default) audio device, writing to an output file is fully optional.
.PP
During playback the stream can be suspened and resumed using the spacebar key.
.SH AUTHOR
Written by Erik Hofman <tech@adalin.com>
.SH SEE ALSO
aaxplaymidi(1), aaxinfo(1), aaxcvt(1)



ChangeLog
* Sat Oct 13 2018 - tech@adalin.org
- AeonWave (3.5.1) stable; urgency=low

 * Improve various instrument configurations.
 * Fix some more issues with the timed gain filter and timed pitch effect.
 * Remove a debugging statement.

-- Erik Hofman <tech@adalin.org>  Sat 13 oct  2018 10:37:00 + 0200

* Thu Sep 27 2018 - tech@adalin.org
- AeonWave (3.5.0) stable; urgency=low

 * Add the option to set the location of the shared-data directory at runtime
   using the new aaxDriverSetSetup function.
 * Add a C++ instrument header file which can be used to simulate a single,
   complete, musical instrument. And use it for the MIDI playback code.
 * Re-initialize the emitter after it is set to suspended. This makes the
   emitter ready for the next playing state.
 * Fix a possible race condition when using tied properties.
 * Fix some issues with the timed gain filter and timed pitch effect.
 * Fix an issue when two registered audio-frames both use a delay effect.
 * Add the option to define the loop staring point and loop end point in AAXS
   files that include a file in the sound section.
 * Allow AAXS defined filters and effects to be set as optional, in which case
   they will be omitted on systems which are low on resources (CPU and memory).
 * AAXS generated buffers will now use mipmapping to extend the pitch range
   beyond a pitch level of 32.

-- Erik Hofman <tech@adalin.org>  Thu 27 Sep  2018 09:13:00 + 0200

* Wed Aug 01 2018 - tech@adalin.org
- AeonWave (3.4.0) stable; urgency=low

 * Add support for the ISO9613-1 specified distance attenuation model.
 * If the ISO9613-1 distance  model is applied, and the velocity effect was not
   yet set, then the corresponding speed of sound will be set automatically.
 * If the source has stopped then finish the current time envelope and stop,
   otherwise an evelope with repeat set to max will never stop until the
   emitter is set to processed.
 * Add a RingModulator effect, mainly to be able to simulate atmospheric
   anomalies for wireless communications. But useful for other purposes too.
 * Remove debugging output from ogg streaming code.

-- Erik Hofman <tech@adalin.org>  Wed 01 Aug  2018 11:00:00 + 0200

* Fri Jul 20 2018 - tech@adalin.org
- AeonWave (3.3.0) stable; urgency=low

 * Automatically convert an effect name like "effects/nature/birds" to a full
   path name using APP_DATA_DIR when it is parsed to the buffer retrieving
   function of the AeonWave class.
 * Add a Param class which encapsulates a float variable and which can
   be tied to a DSP parameter. Every time the param gets a new value
   assigned the DSP parameter will be updated too.
 * Fix a small problem with stopping an emitter which has a timed-gain envelope
   defined with one of the time settings to "inf". They should now finish
   properly after the emitter is requested to stop playing.
 * Overhaul of the C++ classes to make them C++11 compliant.
 * Remove support for an ancient way of handling config files.
 * Prevent a crash when the state of the timed-gain filter or timed-pitch effect
   is not set before it is registered the sensor, audioframe or emitter.

-- Erik Hofman <tech@adalin.org>  Fri 20 Jul  2018 14:01:00 + 0200

* Tue Jul 03 2018 - tech@adalin.org
- AeonWave (3.2.3) stable; urgency=low
 * Fix a missing offset(enum aaxType) error for the AeonWave class.
 * Fix a problem with AAX_SHARED_DATA_DIR support for Windows.

-- Erik Hofman <tech@adalin.org>  Tue 03 Jul  2018 08:49:00 + 0200

* Sat Jun 30 2018 - tech@adalin.org
- AeonWave (3.2.2) stable; urgency=medium
 * Add the AAX_PCM24S_PACKED format for 3-byte packed 24-bit samples.
   AAX_PCM24S is used for 32-bit encoded 24-bit samples.
 * Fix 24-bit packed PCM data handling.
 * Fix a possible stream synchronization problem at the start of streams
   that support ICY metadata song information.

-- Erik Hofman <tech@adalin.org>  Sat 30 Jun  2018 11:06:00 + 0200

* Sat Jun 23 2018 - tech@adalin.org
- AeonWave (3.2.1) stable; urgency=low
 * Improve timing for some backends by using a regular interval method
   instead of waiting just a number of milliseconds from the time of calling.
   This moslty affected OSS and SDL backends.
 * Move the mixing code from the callback to the backend-thread for the SDL
   backend reducing the time spent in the callers code to a mere memcopy.
 * Enable the SDL directsound backend for Windows XP.
 * Rename the angular filter (which was lost in translation) to directional
   filter. For backwards compatibility the angular name is still being
   recognized internally and the library is binary compatible with previous
   versions.

-- Erik Hofman <tech@adalin.org>  Sat 23 Jun  2018 12:31:00 + 0200

* Sat Jun 16 2018 - tech@adalin.org
- AeonWave (3.2.0) stable; urgency=high
 * Add SDL audio support for backends AeonWave does not support natively.
   This potentially opens up the option to use AeonWave on MacOS and BSD.
   SDL 2.0.5 or later is required.
 * Allow retrieving the data directory which AeonWave uses to store shared data
   like preset files by calling aaxDriverGetSetup(cfg, AAX_SHARED_DATA_DIR)
 * Allow defining AAX_ENVELOPE_FOLLOW or AAX_INVERSE_ENVELOPE_FOLLOW alongside
   the waveform type to control the noise behavior when setting the state for
   the bitcrusher filter. The default behavior is still a constant noise level.
 * Fix a number of potential buffer overflows when enumerating the available
   backend interfaces.
 * Fix a number of one-off memory leaks.

-- Erik Hofman <tech@adalin.org>  Sat 16 Jun  2018 13:30:00 + 0200

* Tue Jun 05 2018 - tech@adalin.org
- AeonWave (3.1.2) stable; urgency=low
 * Add the bitcrusher filter to simulate (communication line) sound degradation.
 * Delever on the promis that one can add a named, but empty, buffer in
   case aaxBufferReadFromStream fails when calling buffer() from the
   aeonwave class.

-- Erik Hofman <tech@adalin.org>  Tue 05 Jun  2018 20:48:00 + 0200

* Fri Jun 01 2018 - tech@adalin.org

src/aaxplay.c
/*
 * Copyright (C) 2008-2018 by Erik Hofman.
 * Copyright (C) 2009-2018 by Adalin B.V.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 *    1. Redistributions of source code must retain the above copyright notice,
 *        this list of conditions and the following disclaimer.
 * 
 *    2. Redistributions in binary form must reproduce the above copyright
 *        notice, this list of conditions and the following disclaimer in the
 *        documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY ADALIN B.V. ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN
 * NO EVENT SHALL ADALIN B.V. OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES 
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR 
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUTOF THE USE 
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * The views and conclusions contained in the software and documentation are
 * those of the authors and should not be interpreted as representing official
 * policies, either expressed or implied, of Adalin B.V.
 */

#if HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>
#include <math.h>

#include <aax/aax.h>

#include "base/types.h"
#include "driver.h"
#include "wavfile.h"

#define IFILE_PATH		SRC_PATH"/stereo.mp3"
#define OFILE_PATH		"aaxout.wav"

void
help()
{
    aaxConfig cfgi, cfgo;

    printf("aaxplay version %i.%i.%i\n\n", AAX_UTILS_MAJOR_VERSION,
                                           AAX_UTILS_MINOR_VERSION,
                                           AAX_UTILS_MICRO_VERSION);
    printf("Usage: aaxplay [options]\n");
    printf("Plays audio from a file or from an audio input device.\n");
    printf("Optionally writes the audio to an output file.\n");

    printf("\nOptions:\n");
    printf("  -i, --input <file>\t\tplayback audio from a file\n");
    printf("  -c, --capture <device>\tcapture from an audio device\n");
    printf("  -d, --device <device>\t\tplayback device (default if not specified)\n");
    printf("  -o, --output <file>\t\talso write to an audio file (optional)\n");
    printf("  -b, --batch\t\t\tprocess as fast as possible (Audio Files only)\n");
    printf("  -v, --verbose\t\t\tshow extra playback information\n");
    printf("  -h, --help\t\t\tprint this message and exit\n");
    printf("Either --input or --capture can be used but not both.\n");
    printf("For a list of device names run: aaxinfo\n");

    printf("\nAudio will always be sent to the (default) audio device,\n");
    printf("writing to an output file is fully optional.\n");

    cfgi = aaxDriverGetByName("AeonWave on Audio Files", AAX_MODE_READ);
    cfgo = aaxDriverGetByName("AeonWave on Audio Files", AAX_MODE_WRITE_STEREO);
    if (cfgi || cfgo) {
        printf("\nSupported file formats:\n");
    }

    if (cfgi)
    {
        const char *d, *s;

        d = aaxDriverGetDeviceNameByPos(cfgi, 0, AAX_MODE_READ);
        s = aaxDriverGetInterfaceNameByPos(cfgi, d, 0, AAX_MODE_READ);
        printf("  - input : %s\n", s);
        aaxDriverDestroy(cfgi);
    }

    if (cfgo)
    {
        const char *d, *s;

        d = aaxDriverGetDeviceNameByPos(cfgo, 0, AAX_MODE_WRITE_STEREO);
        s = aaxDriverGetInterfaceNameByPos(cfgo, d, 0, AAX_MODE_WRITE_STEREO);
        printf("  - output: %s\n", s);
        aaxDriverDestroy(cfgo);
    }
    printf("\nUse aaxplaymidi for playing MIDI files.\n");

    printf("\nDuring playback the stream can be suspened and resumed using "
           "the spacebar key.\n");
    printf("\n");

    exit(-1);
}

int main(int argc, char **argv)
{
    char *devname, *idevname;
    char ibuf[256], obuf[256];
    char *infile, *outfile;
    aaxConfig config = NULL;
    aaxConfig record = NULL;
    aaxConfig file = NULL;
    int res, rv = 0;
    int verbose = 0;

    if (argc == 1 || getCommandLineOption(argc, argv, "-h") ||
                     getCommandLineOption(argc, argv, "--help"))
    {
        help();
    }

    if (getCommandLineOption(argc, argv, "-v") || 
        getCommandLineOption(argc, argv, "--verbose"))
    {
       verbose = 1;
    }

    idevname = getCaptureName(argc, argv);
    infile = getInputFile(argc, argv, IFILE_PATH);
    if (!idevname)
    {
       if (infile)
       {
           snprintf(ibuf, 256, "AeonWave on Audio Files: %s", infile);
           idevname = ibuf;
       }
       else {
          help();
       }
    }

    devname = getDeviceName(argc, argv);
    config = aaxDriverOpenByName(devname, AAX_MODE_WRITE_STEREO);
    testForError(config, "Audio output device is not available.");

    if (config)
    {
        record = aaxDriverOpenByName(idevname, AAX_MODE_READ);
        if (!record)
        {
            printf("File not found: %s\n", infile);
            exit(-1);
        }
    }

    outfile = getOutputFile(argc, argv, NULL);
    if (outfile)
    {
        snprintf(obuf, 256, "AeonWave on Audio Files: %s", outfile);
        file = aaxDriverOpenByName(obuf, AAX_MODE_WRITE_STEREO);
    }
    else {
        file = NULL;
    }

    if (config && record && (rv >= 0))
    {
        char batch = getCommandLineOption(argc, argv, "-b") ||
                     getCommandLineOption(argc, argv, "--batch");
        char fparam = getCommandLineOption(argc, argv, "-f") ||
                      getCommandLineOption(argc, argv, "-frame");
        float pitch = getPitch(argc, argv);
        float dhour, hour, minutes, seconds;
        float duration, freq;
        unsigned int max_samples;
        int key, paused;
        aaxFrame frame = NULL;
        aaxEffect effect;
//      aaxFilter filter;
        char tstr[80];
        int state;

        if (batch && !aaxMixerGetSetup(config, AAX_BATCHED_MODE)) {
            printf("Warning: Batched mode not supported for this backend\n");
        }

        /** mixer */
        res = aaxMixerSetSetup(config, AAX_REFRESHRATE, 64);
        testForState(res, "aaxMixerSetSetup");

        res = aaxMixerSetState(config, AAX_INITIALIZED);
        testForState(res, "aaxMixerInit");

        res = aaxMixerSetState(config, AAX_PLAYING);
        testForState(res, "aaxMixerStart");

        if (fparam)
        {
            printf("  using audio-frames\n");

            /** audio frame */
            frame = aaxAudioFrameCreate(config);
            testForError(frame, "Unable to create a new audio frame\n");   

            /** register audio frame */
            res = aaxMixerRegisterAudioFrame(config, frame);
            testForState(res, "aaxMixerRegisterAudioFrame");

            res = aaxAudioFrameSetState(frame, AAX_PLAYING);
            testForState(res, "aaxAudioFrameSetState");

            if (pitch != 1.0f)
            {
#if 0
                effect = aaxAudioFrameGetEffect(frame, AAX_PITCH_EFFECT);
                testForError(effect, "aaxEffectCreate");

                aaxEffectSetParam(effect, AAX_PITCH, AAX_LINEAR, pitch);
#else
                effect = aaxAudioFrameGetEffect(frame,AAX_DYNAMIC_PITCH_EFFECT);
                testForError(effect, "aaxEffectCreate");

                res = aaxEffectSetSlot(effect, 0, AAX_LINEAR,
                                          0.0f, 0.025f, pitch, 0.0f);
                testForState(res, "aaxEffectSetSlot");

                res = aaxEffectSetState(effect, AAX_SINE_WAVE);
                testForState(res, "aaxEffectSetState");
#endif
                res = aaxAudioFrameSetEffect(frame, effect);
                testForState(res, " aaxAudioFrameSetEffect");

                aaxEffectDestroy(effect);
                testForState(res, "aaxEffectDestroy");
            }

            /** sensor */
            res = aaxAudioFrameRegisterSensor(frame, record);
            testForState(res, "aaxAudioFrameRegisterSensor");
        }
        else
        {
#if 0
            effect = aaxMixerGetEffect(config, AAX_PITCH_EFFECT);
            testForError(effect, "aaxEffectCreate");

            aaxEffectSetParam(effect, AAX_PITCH, AAX_LINEAR, pitch);

            res = aaxMixerSetEffect(record, effect);
            testForState(res, "aaxEmitterSetEffect");

            res = aaxEffectDestroy(effect);
            testForState(res, "aaxEffectDestroy");
#endif

            /** sensor */
            res = aaxMixerRegisterSensor(config, record);
            testForState(res, "aaxMixerRegisterSensor");
        }

        if (pitch != 1.0f)
        {
            effect = aaxMixerGetEffect(record, AAX_DYNAMIC_PITCH_EFFECT);
            testForError(effect, "aaxEffectCreate");

            res = aaxEffectSetSlot(effect, 0, AAX_LINEAR,
                                      0.0f, frame ? 0.5f : 0.06f, pitch, 0.0f);
            testForState(res, "aaxEffectSetSlot");

            res = aaxEffectSetState(effect, AAX_TRIANGLE_WAVE);
            testForState(res, "aaxEffectSetState");

            res = aaxMixerSetEffect(record, effect);
            testForState(res, "aaxEmitterSetEffect");

            res = aaxEffectDestroy(effect);
            testForState(res, "aaxEffectDestroy");
        }

        if (file)
        {
            res = aaxMixerRegisterSensor(config, file);
            testForState(res, "aaxMixerRegisterSensor file out");
        }

#if 0
        /** set capturing Auto-Gain Control (AGC): 0dB */
        filter = aaxMixerGetFilter(record, AAX_VOLUME_FILTER);
        if (filter)
        {
            aaxFilterSetParam(filter, AAX_AGC_RESPONSE_RATE, AAX_LINEAR, 1.5f);
            aaxMixerSetFilter(record, filter);
            res = aaxFilterDestroy(filter);
        }
#endif

        /** must be called after aaxMixerRegisterSensor */
        res = aaxMixerSetState(record, AAX_INITIALIZED);
        testForState(res, "aaxMixerSetInitialize");

        res = aaxSensorSetState(record, AAX_CAPTURING);
        testForState(res, "aaxSensorCaptureStart");

        if (verbose)
        {
            unsigned int samples = aaxMixerGetSetup(record, AAX_SAMPLES_MAX);
            if (samples) {
              printf(" Audio format: %i Hz, %i bits/sample, %i tracks, %i samples\n",
                     aaxMixerGetSetup(record, AAX_FREQUENCY),
                     aaxGetBitsPerSample(aaxMixerGetSetup(record, AAX_FORMAT)),
                     aaxMixerGetSetup(record, AAX_TRACKS), samples);
           } else {
              printf(" Audio format: %i Hz, %i bits/sample, %i tracks\n",
                     aaxMixerGetSetup(record, AAX_FREQUENCY),
                     aaxGetBitsPerSample(aaxMixerGetSetup(record, AAX_FORMAT)),
                     aaxMixerGetSetup(record, AAX_TRACKS));
           }
        }

        if (record && verbose)
        {
            const char *s;

            s = aaxDriverGetSetup(record, AAX_MUSIC_PERFORMER_STRING);
            if (s) printf(" Performer: %s\n", s);

            s = aaxDriverGetSetup(record, AAX_TRACK_TITLE_STRING);
            if (s) printf(" Title    : %s\n", s);

            s = aaxDriverGetSetup(record, AAX_ALBUM_NAME_STRING);
            if (s) printf(" Album    : %s\n", s);

            s = aaxDriverGetSetup(record, AAX_SONG_COMPOSER_STRING);
            if (s) printf(" Composer : %s\n", s);

            s = aaxDriverGetSetup(record, AAX_ORIGINAL_PERFORMER_STRING);
            if (s) printf(" Original : %s\n", s);

            s = aaxDriverGetSetup(record, AAX_MUSIC_GENRE_STRING);
            if (s) printf(" Genre    : %s\n", s);
            s = aaxDriverGetSetup(record, AAX_RELEASE_DATE_STRING);
            if (s) printf(" Release date: %s\n", s);

            s = aaxDriverGetSetup(record, AAX_TRACK_NUMBER_STRING);
            if (s) printf(" Track number: %s\n", s);

            s = aaxDriverGetSetup(record, AAX_SONG_COPYRIGHT_STRING);
            if (s) printf(" Copyright:  %s\n", s);

            s = aaxDriverGetSetup(record, AAX_WEBSITE_STRING);
            if (s) printf(" Website  : %s\n", s);
        }

        if (file)
        {
            res = aaxMixerSetState(file, AAX_INITIALIZED);
            testForState(res, "aaxMixerSetInitialize");

            res = aaxMixerSetState(file, AAX_PLAYING);
            testForState(res, "aaxSensorCaptureStart");
        }

        set_mode(1);
        freq = (float)aaxMixerGetSetup(record, AAX_FREQUENCY);
        max_samples = aaxMixerGetSetup(record, AAX_SAMPLES_MAX);
        if (max_samples)
        {
            duration = (float)max_samples/freq;
            seconds = duration;
            dhour = floorf(seconds/(60.0f*60.0f));
            seconds -= dhour*60.0f*60.0f;
            minutes = floorf(seconds/60.0f);
            seconds -= minutes*60.0f;
            if (dhour) {
               snprintf(tstr, 80, "%s  %02.0f:%02.0f:%02.0f %s\r",
                                  "pos: % 5.1f (%02.0f:%02.0f:%04.1f) of ",
                                  dhour, minutes, seconds, " % 3.0f%");
            } else {
               snprintf(tstr, 80, "%s  %02.0f:%02.0f %s\r",
                                  "pos: % 5.1f (%02.0f:%04.1f) of ",
                                  minutes, seconds, " % 3.0f%");
           }
        }
        else
        {
           dhour = duration = AAX_FPINFINITE;
           snprintf(tstr, 80, "%s\r", "pos: % 5.1f (%02.0f:%02.0f:%04.1f)");
        }

        paused = AAX_FALSE;
        do
        {
            if (verbose)
            {
                float pos = (float)aaxSensorGetOffset(record, AAX_SAMPLES)/freq;
                const char *p, *t;

                p = aaxDriverGetSetup(record, AAX_MUSIC_PERFORMER_UPDATE);
                t = aaxDriverGetSetup(record, AAX_TRACK_TITLE_UPDATE);
                if (p && t) {
                    printf("\r\033[K Playing  : %s - %s\n", p, t);
                } else if (p) {
                    printf("\r\033[K Performer: %s\n", p);
                } else if (t) {
                    printf("\r\033[K Title    : %s\n", t);
                }

                if (duration != AAX_FPINFINITE)
                {
                    seconds = duration-pos;
                    hour = floorf(seconds/(60.0f*60.0f));
                    seconds -= hour*60.0f*60.0f;
                    minutes = floorf(seconds/60.0f);
                    seconds -= minutes*60.0f;
                    if (dhour) {
                       printf(tstr, pos, hour, minutes, seconds, 100*pos/duration);
                    } else {
                       printf(tstr, pos, minutes, seconds, 100*pos/duration);
                    }
                }
                else
                {
                    seconds = pos;
                    hour = floorf(seconds/(60.0f*60.0f));
                    seconds -= hour*60.0f*60.0f;
                    minutes = floorf(seconds/60.0f);
                    seconds -= minutes*60.0f;
                    printf(tstr, pos, hour, minutes, seconds);
                }
                fflush(stdout);
            }

            key = get_key();
            if (key)
            {
               if (key == ' ')
               {
                  if (paused)
                  {
                     aaxMixerSetState(config, AAX_PLAYING);
                     printf("\nRestart playback.\n");
                     paused = AAX_FALSE;
                  }
                  else
                  {
                     aaxMixerSetState(config, AAX_SUSPENDED);
                     printf("\nPause playback.\n");
                     paused = AAX_TRUE;
                  }
               }
               else {
                  break;
               }
            }

            if (batch) {
               res = aaxMixerSetState(config, AAX_UPDATE);
            } else {
                msecSleep(250);
            }
            state = aaxMixerGetState(record);
        }
        while (state == AAX_PLAYING);
        printf("\n");
        set_mode(0);

        res = aaxMixerSetState(config, AAX_STOPPED);
        testForState(res, "aaxMixerSetState");

        if (frame)
        {
            res = aaxAudioFrameSetState(frame, AAX_STOPPED);
            testForState(res, "aaxAudioFrameSetState");
        }

        res = aaxSensorSetState(record, AAX_STOPPED);
        testForState(res, "aaxSensorCaptureStop");

        if (frame)
        {
            res = aaxAudioFrameDeregisterSensor(frame, record);
            testForState(res, "aaxAudioFrameDeregisterSensor");

            res = aaxMixerDeregisterAudioFrame(config, frame);
            testForState(res, "aaxMixerDeregisterAudioFrame");

            res = aaxAudioFrameDestroy(frame);
            testForState(res, "aaxAudioFrameDestroy");
        }
        else
        {
            res = aaxMixerDeregisterSensor(config, record);
            testForState(res, "aaxMixerDeregisterSensor");
        }
    }
    else {
        printf("Unable to open capture device.\n");
    }

    if (record)
    {
        res = aaxDriverClose(record);
        res = aaxDriverDestroy(record);
    }

    if (file)
    {
        res = aaxMixerSetState(file, AAX_STOPPED);
        testForState(res, "aaxMixerSetState");

        res = aaxMixerDeregisterSensor(config, file);
        testForState(res, "aaxMixerDeregisterSensor file out");

        res = aaxDriverClose(file);
        testForState(res, "aaxDriverClose");

        res = aaxDriverDestroy(file);
        testForState(res, "aaxDriverDestroy");
    }

    res = aaxDriverClose(config);
    testForState(res, "aaxDriverClose");

    res = aaxDriverDestroy(config);
    testForState(res, "aaxDriverDestroy");

    return rv;
}

